### ISSUE TITLE
No warning reported when storing address of the local variable (Wdangling-pointer)

### ISSUE THREAD

**Issue by mushenoy** [llvm/llvm-project#83719]
Clang seem to miss detecting the dangling pointer warning (Wdangling-pointer) found with GCC

Sample Program:
```
#include <stdio.h>
void test (int **p)
{
  int x = 7;
  *p = &x;
}
int main ()
{
  int *xp = NULL;
  test (&xp);

  printf ("dereference of xp = %d\n", *xp);
  return 0;
}
```

Compilation using clang:
No warnings reported
```
# clang -Wall -c dangling.c
#
```

Compilation using gcc:
```
# gcc -Wall -c  dangling.c
dangling.c: In function test:
dangling.c:5:6: warning: storing the address of local variable x in *p [-Wdangling-pointer=]
    5 |   *p = &x;
      |   ~~~^~~~
dangling.c:4:7: note: x declared here
    4 |   int x = 7;
      |       ^
dangling.c:2:18: note: p declared here
    2 | void test (int **p)
      |
```

Compiler versions: 
``` 
clang version 16.0.6 
gcc (GCC) 13.2.1
```

**Comment by shafik** [comment]
May be related to: https://github.com/llvm/llvm-project/issues/63310

**Comment by dcb314** [comment]
A build of recent clang trunk with gcc suggests there are about
30 different places in clang trunk source code where this warning
is produced.

 Here are the most popular 8:

     20 /home/dcb40b/llvm/trunk/clang/lib/AST/ByteCode/Interp.h:550:14: warning: possibly dangling reference to a temporary [-Wdangling-reference]
     20 /home/dcb40b/llvm/trunk/clang/lib/AST/ByteCode/Interp.h:549:14: warning: possibly dangling reference to a temporary [-Wdangling-reference]
     20 /home/dcb40b/llvm/trunk/clang/lib/AST/ByteCode/Interp.h:548:14: warning: possibly dangling reference to a temporary [-Wdangling-reference]
     20 /home/dcb40b/llvm/trunk/clang/lib/AST/ByteCode/Interp.h:547:14: warning: possibly dangling reference to a temporary [-Wdangling-reference]
     20 /home/dcb40b/llvm/trunk/clang/lib/AST/ByteCode/Interp.h:495:14: warning: possibly dangling reference to a temporary [-Wdangling-reference]
     20 /home/dcb40b/llvm/trunk/clang/lib/AST/ByteCode/Interp.h:494:14: warning: possibly dangling reference to a temporary [-Wdangling-reference]
     20 /home/dcb40b/llvm/trunk/clang/lib/AST/ByteCode/Interp.h:493:14: warning: possibly dangling reference to a temporary [-Wdangling-reference]
     20 /home/dcb40b/llvm/trunk/clang/lib/AST/ByteCode/Interp.h:492:14: warning: possibly dangling reference to a temporary [-Wdangling-reference]

The count at the start of the line indicates how many times 
the warning is produced in the build.

Here are the next 7:

      8 /home/dcb40b/llvm/trunk/llvm/lib/Transforms/IPO/AttributorAttributes.cpp:450:23: warning: possibly dangling reference to a temporary [-Wdangling-reference]
      5 /home/dcb40b/llvm/trunk/llvm/include/llvm/CodeGen/SelectionDAG.h:323:27: warning: storing the address of local variable ‘Listener’ in ‘*this.llvm::SelectionDAG::UpdateListeners’ [-Wdangling-pointer=]
      4 /home/dcb40b/llvm/trunk/llvm/lib/Object/ELFObjectFile.cpp:954:3: warning: possibly dangling reference to a temporary [-Wdangling-reference]
      4 /home/dcb40b/llvm/trunk/llvm/lib/Object/ELFObjectFile.cpp:898:15: warning: possibly dangling reference to a temporary [-Wdangling-reference]
      4 /home/dcb40b/llvm/trunk/llvm/lib/Object/ELF.cpp:961:3: warning: possibly dangling reference to a temporary [-Wdangling-reference]
      3 /home/dcb40b/llvm/trunk/llvm/lib/ObjectYAML/WasmEmitter.cpp:332:70: warning: possibly dangling reference to a temporary [-Wdangling-reference]
      2 /home/dcb40b/llvm/trunk/clang/include/clang/Sema/Template.h:438:41: warning: storing the address of local variable ‘InstScope’ in ‘*this.clang::Sema::CurrentInstantiationScope’ [-Wdangling-pointer=]

The remainder are singletons:

      1 /home/dcb40b/llvm/trunk/llvm/utils/TableGen/GlobalISelCombinerEmitter.cpp:1910:15: warning: possibly dangling reference to a temporary [-Wdangling-reference]
      1 /home/dcb40b/llvm/trunk/llvm/lib/Transforms/Scalar/ADCE.cpp:303:72: warning: possibly dangling reference to a temporary [-Wdangling-reference]
      1 /home/dcb40b/llvm/trunk/llvm/lib/Transforms/IPO/AttributorAttributes.cpp:6384:25: warning: possibly dangling reference to a temporary [-Wdangling-reference]
      1 /home/dcb40b/llvm/trunk/llvm/lib/DebugInfo/PDB/Native/NativeSession.cpp:392:9: warning: possibly dangling reference to a temporary [-Wdangling-reference]
      1 /home/dcb40b/llvm/trunk/llvm/lib/DebugInfo/PDB/Native/NativeSession.cpp:239:11: warning: possibly dangling reference to a temporary [-Wdangling-reference]
      1 /home/dcb40b/llvm/trunk/llvm/include/llvm/CodeGen/SelectionDAG.h:517:53: warning: storing the address of local variable ‘FlagsInserter’ in ‘*DAG.llvm::SelectionDAG::Inserter’ [-Wdangling-pointer=]
      1 /home/dcb40b/llvm/trunk/llvm/include/llvm/Analysis/RegionInfoImpl.h:455:3: warning: possibly dangling reference to a temporary [-Wdangling-reference]
      1 /home/dcb40b/llvm/trunk/llvm/include/llvm/Analysis/RegionInfoImpl.h:444:5: warning: possibly dangling reference to a temporary [-Wdangling-reference]
      1 /home/dcb40b/llvm/trunk/clang/lib/StaticAnalyzer/Core/Store.cpp:522:23: warning: possibly dangling reference to a temporary [-Wdangling-reference]
      1 /home/dcb40b/llvm/trunk/clang/lib/StaticAnalyzer/Core/Store.cpp:508:23: warning: possibly dangling reference to a temporary [-Wdangling-reference]
      1 /home/dcb40b/llvm/trunk/clang/lib/StaticAnalyzer/Core/SimpleConstraintManager.cpp:123:25: warning: possibly dangling reference to a temporary [-Wdangling-reference]
      1 /home/dcb40b/llvm/trunk/clang/lib/StaticAnalyzer/Checkers/MPI-Checker/MPIChecker.cpp:167:25: warning: possibly dangling reference to a temporary [-Wdangling-reference]
      1 /home/dcb40b/llvm/trunk/clang/lib/AST/ByteCode/EvaluationResult.cpp:207:22: warning: possibly dangling reference to a temporary [-Wdangling-reference]
      1 /home/dcb40b/llvm/trunk/clang/lib/AST/ByteCode/Compiler.h:449:19: warning: storing the address of local variable ‘Scope’ in ‘*this.clang::interp::Compiler<clang::interp::EvalEmitter>::VarScope’ [-Wdangling-pointer=]
      1 /home/dcb40b/llvm/trunk/clang/lib/AST/ByteCode/Compiler.h:449:19: warning: storing the address of local variable ‘Scope’ in ‘*this.clang::interp::Compiler<clang::interp::ByteCodeEmitter>::VarScope’ [-Wdangling-pointer=]

It might be worth fixing up the source code of the clang compiler.

**Comment by shafik** [comment]
CC @tbaederr @balazske since I see bytecode interpreter and static analysis several times in the list above. Worth looking into maybe.

**Comment by zufuliu** [comment]
CC @steakhal, could static analyzer's [StackAddressEscape](https://clang.llvm.org/docs/analyzer/checkers.html#core-stackaddressescape-c) checker be updated to warn [dangling pointer](https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wdangling-pointer) caught by GCC?

**Comment by steakhal** [comment]
I think these are caught. You should use the --analyze flag to enable the Clang Static Analyzer from the command line.

Expect a huge slowdown though.

**Comment by zufuliu** [comment]
Not all caught, https://godbolt.org/z/aTvj83zr4

**Comment by steakhal** [comment]
> Not all caught, https://godbolt.org/z/aTvj83zr4

I only tried the other libked issue because that had a godbolt link to modify. That would be caught.

In your case I don't see undefined behavior that we should catch.

**Comment by zufuliu** [comment]
> In your case I don't see undefined behavior that we should catch.

The lifetime of `Foo inner` is ended at end of inner scope (`inner.ptr` contains garbage), `use2(ptr)` is called with destructed object pointer?

```c++
void foo(bool flag) {
    Foo *ptr = nullptr;
    if (flag) {
        Foo inner;
        ptr = &inner;
        use1(ptr);
    }
    use2(ptr);
}
```

**Comment by steakhal** [comment]
use2 may not use ptr.

**Comment by zufuliu** [comment]
> use2 may not use ptr.

You are correct, when `use2` uses `ptr`, the bug in `foo()` will be caught by [NewDelete](https://clang.llvm.org/docs/analyzer/checkers.html#cplusplus-newdelete) checker. so there is miss detection when body for `use2` is unavailable (e.g. system API).

Updated test for `bar()`: https://godbolt.org/z/GYKrz9b63
```c++
extern void use1(char *ptr);
extern void use2(char *ptr);
extern void use3(char *ptr);
void bar(int flag) {
    char *ptr = nullptr;
    if (flag & 1) {
        char inner1[16];
        ptr = inner1;
        use1(ptr);
    }
    if (flag & 2) {
        char inner2[16];
        use2(inner2);
    }
    use3(ptr);
}
```

`use3(ptr)` is called with pointer to garbage due to stack reuse (like clang). so should be caught?

**Comment by steakhal** [comment]
I agree that we should in this case. Here we have a pretty clear intent when passig it as a direct argument.
But for instance if we had a struct (like optional) that had inside a flag encoding that it holds a valid pointer or not, then passing a struct like that and raising a warning could be considered as a FP.
So the warning is justified in this specific case but not in general.

**Comment by steakhal** [comment]
If only we could know that use3 will unconditionally dereference this pointer, then it would be clear.

**Comment by zufuliu** [comment]
I was thinking [StackAddressEscape](https://clang.llvm.org/docs/analyzer/checkers.html#core-stackaddressescape-c) could be extended to check addresses to stack object escape its lifetime/enclosing scope, than just function scope currently implemented.

> Check that addresses to stack memory do not escape the function.

compared to GCC's [`-Wdangling-pointer`](https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wdangling-pointer):
> Warn about uses of pointers (or C++ references) to objects with automatic storage duration after their lifetime has ended. This includes local variables declared in nested blocks, compound literals and other unnamed temporary objects. In addition, warn about storing the address of such objects in escaped pointers.

**Comment by steakhal** [comment]
> I was thinking [StackAddressEscape](https://clang.llvm.org/docs/analyzer/checkers.html#core-stackaddressescape-c) could be extended to check addresses to stack object escape its lifetime/enclosing scope, than just function scope currently implemented.
> 
> > Check that addresses to stack memory do not escape the function.

I wish it was easy. It's possible, but it's not trivial to implement.
